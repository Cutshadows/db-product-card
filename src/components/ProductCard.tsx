import { createContext, Fragment} from 'react';
import styles from '../styles/styles.module.css';
import { useProduct } from '../hooks/useProduct';
import { ProductCardProps } from '../interface/Product.dto';
import React from 'react';



export const ProductContext = createContext({ ProductState: { counter: 0, product: { id: '', title: '', img: 'https://w7.pngwing.com/pngs/364/982/png-transparent-python-programming-language-computer-programming-language-angle-text-computer-thumbnail.png' },  maxCount: 10, increaseBy: ( _value: number ):void => {} } });
const { Provider } = ProductContext;










export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: ProductCardProps) => {

	const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct( {onChange, product, value, initialValues} );

	return (
		<Fragment>
			<Provider value={{
				ProductState:{
					counter,
					maxCount,
					increaseBy,
					product,
				}
			}}>
				<div className={`${styles.productCard } ${className}`} style={style}>
					{ children({
						count: counter,
						isMaxCountReached,
						maxCount: initialValues&&initialValues.maxCount,
						product,
						increaseBy,
						reset

					}) }
				</div>
			</Provider>
		</Fragment>
	);
};


