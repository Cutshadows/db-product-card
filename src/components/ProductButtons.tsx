import { useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';
import React from 'react';

export interface IProps {
	className?: string;
	style?: React.CSSProperties;
}

export const ProductButtons = ({className, style}: IProps) => {
	const {ProductState}= useContext(ProductContext);
	const{increaseBy, counter, maxCount} = ProductState;

	const isMaxReached=useCallback(
	  () => !!maxCount && counter === maxCount, 
	  [maxCount, counter]
	);
	

	return (
		<div className={`${styles.buttonsContainer} ${className}` } style={style}>
			<button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
				-
			</button>
			<div className={styles.countLabel}> {counter} </div>
			<button className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`} onClick={() => increaseBy(1)}>
				+
			</button>
		</div>
	);
};