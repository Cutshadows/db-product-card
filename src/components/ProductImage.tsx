import { useContext } from 'react';
import styles from '../styles/styles.module.css';
import imgPython from '../assets/scope_python.webp';
import { ProductContext } from './ProductCard';
import React from 'react';

export interface Props {
	img?: string; 
	className?: string;
	style ?: React.CSSProperties;
}

export const ProductImage = ({ img, className, style }: Props) : JSX.Element => {
	const {ProductState} = useContext(ProductContext);
	const {product} = ProductState;
	
	return (
		<img style={style} src={img || product!.img ||imgPython} alt={img} className={`${styles.productImg} ${className}`} />
	);
};