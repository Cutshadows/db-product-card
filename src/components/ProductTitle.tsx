import { useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';
import React from 'react';

export interface Props { 
	title: string; 
	className?:string;
	style?: React.CSSProperties; 
}


export const ProductTitle = ({ title, className, style }: Props) =>{
	const {ProductState} = useContext(ProductContext);
	const {product} = ProductState;


	return(
		<span className={`${styles.productDescription} ${className}`} style={style}>{!title?product.title : title}</span>
	);
};