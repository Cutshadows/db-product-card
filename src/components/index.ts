import { ProductCard as ProductCardHOC } from './ProductCard';

import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';
import { ProductCardHOCProps } from '../interface/Product.dto';

export { ProductButtons } from './ProductButtons';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';



export const ProductCard: ProductCardHOCProps = Object.assign( ProductCardHOC, {
	Title: ProductTitle,
	Buttons: ProductButtons,
	Image: ProductImage
});

export default ProductCard;