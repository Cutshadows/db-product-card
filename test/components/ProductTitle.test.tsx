import React from 'react';
// import TestRenderer from 'react-test-renderer';
import * as renderer from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../../src/components/';
import { product1, product2 } from '../data/products';

const products = [product1, product2];
describe('ProductTitle', () => {
  test('should view the component with custom title', () => {

        const wrapper = renderer.create(
            <ProductTitle  title="Custom Props" className='custom-class' />
        ).toJSON();

        expect(wrapper).toMatchSnapshot();
  });

  test('should view the components with the name of product', () => {
    const wrapper = renderer.create(
        <ProductCard product={products[1]} >
            {
                ()=> (
                    <ProductTitle title={''} />
                )
            }
        </ProductCard>
    ).toJSON();
    expect(wrapper).toMatchSnapshot();
   });
});
