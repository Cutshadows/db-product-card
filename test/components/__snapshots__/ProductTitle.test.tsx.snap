// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle should view the component with custom title 1`] = `
<span
  className="productDescription custom-class"
>
  Custom Props
</span>
`;

exports[`ProductTitle should view the components with the name of product 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Producto test 2 undefined image
  </span>
</div>
`;
